File: backend/app/schemas/testcase.py
"""Testcase Schemas.

Pydantic schemas for testcase management module.
"""

import json
from datetime import datetime
from typing import Optional, List, Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.testcase import CaseType, Platform, Priority, CaseStatus


# ============ TestStep Schema ============

class TestStep(BaseModel):
    """Represents a single test step.

    A test step contains a description of the action and the expected result.
    """
    step: str = Field(..., description="Step description")
    expected: str = Field(..., description="Expected result")


# ============ TestCaseGroup Schemas ============

class TestCaseGroupBase(BaseModel):
    """Base schema for TestCaseGroup."""
    name: str = Field(..., min_length=1, max_length=100, description="Group name")
    parent_id: Optional[int] = Field(None, description="Parent group ID")
    order: int = Field(0, ge=0, description="Sort order")
    description: Optional[str] = Field(None, description="Group description")


class TestCaseGroupCreate(TestCaseGroupBase):
    """Schema for creating a new TestCaseGroup."""
    pass


class TestCaseGroupUpdate(BaseModel):
    """Schema for updating an existing TestCaseGroup."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    parent_id: Optional[int] = None
    order: Optional[int] = Field(None, ge=0)
    description: Optional[str] = None


class TestCaseGroupResponse(TestCaseGroupBase):
    """Response schema for TestCaseGroup."""
    id: int
    created_by: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_at: datetime
    children: List["TestCaseGroupResponse"] = []
    case_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class TestCaseGroupListResponse(BaseModel):
    """List response for TestCaseGroup."""
    total: int
    items: List[TestCaseGroupResponse]


# ============ TestCase Schemas ============

class TestCaseBase(BaseModel):
    """Base schema for TestCase - 定义见设计文档 4.1"""
    title: str = Field(..., min_length=1, max_length=200, description="用例标题")
    case_type: Optional[CaseType] = Field(None, description="用例类型")
    platform: Optional[Platform] = Field(None, description="所属平台")
    priority: Optional[Priority] = Field(None, description="重要程度")
    is_core: Optional[bool] = Field(False, description="核心用例")
    owner: Optional[str] = Field(None, max_length=50, description="维护人")
    developer: Optional[str] = Field(None, max_length=50, description="开发负责人")
    page_url: Optional[str] = Field(None, max_length=500, description="页面地址")
    preconditions: Optional[str] = Field(None, description="前置条件（富文本）")
    steps: Optional[Any] = Field(None, description="测试步骤（表格）")
    remark: Optional[str] = Field(None, description="备注（富文本）")
    tags: Optional[Any] = Field(None, description="标签数组")
    status: Optional[CaseStatus] = Field(CaseStatus.DRAFT, description="用例状态")


class TestCaseCreate(TestCaseBase):
    """Schema for creating a new TestCase."""
    group_id: int = Field(..., description="Group ID")


class TestCaseUpdate(BaseModel):
    """Schema for updating an existing TestCase."""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    group_id: Optional[int] = None
    case_type: Optional[CaseType] = None
    platform: Optional[Platform] = None
    priority: Optional[Priority] = None
    status: Optional[CaseStatus] = None
    is_core: Optional[bool] = None
    owner: Optional[str] = Field(None, max_length=50)
    developer: Optional[str] = Field(None, max_length=50)
    page_url: Optional[str] = Field(None, max_length=500)
    preconditions: Optional[str] = None
    steps: Optional[Any] = None
    remark: Optional[str] = None
    tags: Optional[Any] = None


class TestCaseResponse(BaseModel):
    """Response schema for TestCase."""
    id: int
    group_id: int
    code: str
    order: int = 0
    title: str
    case_type: Optional[CaseType] = None
    platform: Optional[Platform] = None
    priority: Optional[Priority] = None
    is_core: bool = False
    owner: Optional[str] = None
    developer: Optional[str] = None
    page_url: Optional[str] = None
    preconditions: Optional[str] = None
    steps: Optional[Any] = None
    remark: Optional[str] = None
    tags: Optional[Any] = None
    status: CaseStatus = CaseStatus.DRAFT
    created_by: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator('tags', 'steps', mode='before')
    @classmethod
    def parse_json_fields(cls, v):
        """Parse JSON string fields to objects."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (json.JSONDecodeError, ValueError):
                return v
        return v


class TestCaseListResponse(BaseModel):
    """List response for TestCase."""
    total: int
    items: List[TestCaseResponse]


class TestCaseDetailResponse(TestCaseResponse):
    """Detailed response schema for TestCase including attachments, comments, and history."""
    attachments: List["TestCaseAttachmentResponse"] = []
    comments: List["TestCaseCommentResponse"] = []
    history: List["TestCaseHistoryResponse"] = []


# ============ TestCaseAttachment Schemas ============

class TestCaseAttachmentResponse(BaseModel):
    """Response schema for TestCaseAttachment."""
    id: int
    case_id: int
    filename: str
    file_path: str
    file_size: int = 0
    mime_type: Optional[str] = None
    uploaded_by: Optional[str] = None
    uploaded_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TestCaseAttachmentCreate(BaseModel):
    """Schema for creating a new TestCaseAttachment."""
    filename: str = Field(..., max_length=255)
    file_path: str = Field(..., max_length=500)
    file_size: int = Field(0, ge=0)
    mime_type: Optional[str] = Field(None, max_length=100)


# ============ TestCaseComment Schemas ============

class TestCaseCommentBase(BaseModel):
    """Base schema for TestCaseComment."""
    content: str = Field(..., min_length=1, description="Comment content")


class TestCaseCommentCreate(TestCaseCommentBase):
    """Schema for creating a new TestCaseComment."""
    case_id: int
    parent_id: Optional[int] = Field(None, description="Parent comment ID for replies")


class TestCaseCommentUpdate(BaseModel):
    """Schema for updating an existing TestCaseComment."""
    content: str = Field(..., min_length=1)


class TestCaseCommentResponse(TestCaseCommentBase):
    """Response schema for TestCaseComment."""
    id: int
    case_id: int
    author: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    replies: List["TestCaseCommentResponse"] = []

    model_config = ConfigDict(from_attributes=True)


# ============ TestCaseHistory Schema ============

class TestCaseHistoryResponse(BaseModel):
    """Response schema for TestCaseHistory."""
    id: int
    case_id: int
    action: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    operator: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator('old_value', 'new_value', mode='before')
    @classmethod
    def parse_json_fields(cls, v):
        """Parse JSON string fields to objects."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (json.JSONDecodeError, ValueError):
                return v
        return v


# ============ TreeNode Schema ============

class TreeNode(BaseModel):
    """Schema for tree node display.

    Used for representing the hierarchical structure of testcase groups.
    """
    id: int
    label: str
    parent_id: Optional[int] = None
    order: int = 0
    case_count: int = 0
    children: List["TreeNode"] = []

    model_config = ConfigDict(from_attributes=True)


# ============ Batch Operation Schemas ============

class BatchMoveRequest(BaseModel):
    """Schema for batch move operation."""
    case_ids: List[int] = Field(..., min_length=1, description="List of case IDs to move")
    target_group_id: int = Field(..., description="Target group ID")


class BatchDeleteRequest(BaseModel):
    """Schema for batch delete operation."""
    case_ids: List[int] = Field(..., min_length=1, description="List of case IDs to delete")


class BatchCopyRequest(BaseModel):
    """Schema for batch copy operation."""
    case_ids: List[int] = Field(..., min_length=1, description="List of case IDs to copy")
    target_group_id: int = Field(..., description="Target group ID")


class BatchUpdateStatusRequest(BaseModel):
    """Schema for batch update status operation."""
    case_ids: List[int] = Field(..., min_length=1, description="List of case IDs to update")
    status: CaseStatus = Field(..., description="New status")


class BatchOperationResponse(BaseModel):
    """Response schema for batch operations."""
    success: bool
    affected_count: int
    message: str


# ============ Forward Reference Resolution ============

# Rebuild models to resolve forward references
TestCaseGroupResponse.model_rebuild()
TestCaseDetailResponse.model_rebuild()
TestCaseCommentResponse.model_rebuild()
TreeNode.model_rebuild()

File: frontend/src/api/testcase.ts
import axios from 'axios'

const api = axios.create({
  baseURL: '/api/v1',
  timeout: 10000
})

// =====================
// Enum Constants (设计文档 4.2 & 5.2)
// =====================

// 用例类型 - 严格按照设计文档 4.2
export const CASE_TYPES = [
  { value: '功能测试', label: '功能测试' },
  { value: '性能测试', label: '性能测试' },
  { value: '安全测试', label: '安全测试' },
  { value: '兼容性测试', label: '兼容性测试' },
  { value: '用户体验测试', label: '用户体验测试' },
  { value: '其他', label: '其他' },
] as const

// 所属平台 - 严格按照设计文档 4.2
export const PLATFORMS = [
  { value: 'RN', label: 'RN (Android+H5)' },
  { value: '服务端', label: '服务端' },
  { value: '小程序', label: '小程序' },
  { value: 'Web', label: 'Web' },
  { value: 'H5', label: 'H5' },
] as const

// 重要程度 - 严格按照设计文档 4.2
export const PRIORITIES = [
  { value: 'P0', label: 'P0 (最高)' },
  { value: 'P1', label: 'P1' },
  { value: 'P2', label: 'P2' },
  { value: 'P3', label: 'P3' },
  { value: 'P4', label: 'P4 (最低)' },
] as const

// 用例状态 - 严格按照设计文档 5.2
export const CASE_STATUSES = [
  { value: '草稿', label: '草稿' },
  { value: '评审中', label: '评审中' },
  { value: '通过', label: '通过' },
  { value: '废弃', label: '废弃' },
] as const

// =====================
// Type Definitions
// =====================

// Tree node for group hierarchy
export interface TreeNode {
  id: number
  label: string
  parent_id: number | null
  order: number
  case_count: number
  children: TreeNode[]
}

// Test step within a test case
export interface TestStep {
  step: string
  expected: string
}

// Test case group
export interface TestCaseGroup {
  id: number
  name: string
  parent_id: number | null
  order: number
  description?: string
  created_by: string
  created_at: string
  updated_by: string
  updated_at: string
  case_count: number
}

// Test case
export interface TestCase {
  id: number
  group_id: number
  code: string
  order: number
  title: string
  case_type?: string
  platform?: string
  priority?: string
  is_core: boolean
  owner?: string
  developer?: string
  page_url?: string
  preconditions?: string
  steps?: TestStep[]
  remark?: string
  tags?: string[]
  status: string
  created_by?: string
  created_at: string
  updated_by?: string
  updated_at: string
}

// Test case detail with attachments, comments, and history
export interface TestCaseDetail extends TestCase {
  attachments: TestCaseAttachment[]
  comments: TestCaseComment[]
  history: TestCaseHistory[]
}

// Attachment
export interface TestCaseAttachment {
  id: number
  case_id: number
  filename: string
  file_path: string
  file_size: number
  file_type: string
  uploaded_by: string
  uploaded_at: string
}

// Comment
export interface TestCaseComment {
  id: number
  case_id: number
  parent_id: number | null
  content: string
  created_by: string
  created_at: string
  replies: TestCaseComment[]
}

// Change history
export interface TestCaseHistory {
  id: number
  case_id: number
  field_name: string
  old_value: string | null
  new_value: string | null
  changed_by: string
  changed_at: string
}

// Response type for case list
export interface TestCaseListResponse {
  total: number
  items: TestCase[]
}

// =====================
// API Methods
// =====================

export const testcaseApi = {
  // Group operations
  getGroupTree: () =>
    api.get<TreeNode[]>('/testcase/groups/tree'),

  getGroup: (id: number) =>
    api.get<TestCaseGroup>(`/testcase/groups/${id}`),

  createGroup: (data: Partial<TestCaseGroup>) =>
    api.post<TestCaseGroup>('/testcase/groups', data),

  updateGroup: (id: number, data: Partial<TestCaseGroup>) =>
    api.put<TestCaseGroup>(`/testcase/groups/${id}`, data),

  deleteGroup: (id: number) =>
    api.delete(`/testcase/groups/${id}`),

  // Case operations
  getCases: (groupId: number, skip = 0, limit = 100, keyword?: string) =>
    api.get<TestCaseListResponse>('/testcase/cases', {
      params: { group_id: groupId, skip, limit, keyword }
    }),

  getCase: (id: number) =>
    api.get<TestCaseDetail>(`/testcase/cases/${id}`),

  createCase: (data: Partial<TestCase>) =>
    api.post<TestCase>('/testcase/cases', data),

  updateCase: (id: number, data: Partial<TestCase>) =>
    api.put<TestCase>(`/testcase/cases/${id}`, data),

  deleteCase: (id: number) =>
    api.delete(`/testcase/cases/${id}`),

  copyCase: (id: number) =>
    api.post<TestCase>(`/testcase/cases/${id}/copy`),

  moveCase: (id: number, newGroupId: number) =>
    api.post(`/testcase/cases/${id}/move`, null, {
      params: { new_group_id: newGroupId }
    }),

  reorderCases: (groupId: number, orders: { id: number; order: number }[]) =>
    api.put('/testcase/cases/reorder', orders, {
      params: { group_id: groupId }
    }),

  // Comment operations
  getComments: (caseId: number) =>
    api.get<TestCaseComment[]>(`/testcase/cases/${caseId}/comments`),

  addComment: (caseId: number, data: { content: string; parent_id?: number }) =>
    api.post<TestCaseComment>(`/testcase/cases/${caseId}/comments`, data),

  deleteComment: (id: number) =>
    api.delete(`/testcase/comments/${id}`),

  // Attachment operations
  getAttachments: (caseId: number) =>
    api.get<TestCaseAttachment[]>(`/testcase/cases/${caseId}/attachments`),

  getAttachmentDownloadUrl: (attachmentId: number) =>
    `/api/v1/testcase/attachments/${attachmentId}/download`,

  deleteAttachment: (attachmentId: number) =>
    api.delete(`/testcase/attachments/${attachmentId}`),

  // Import/Export operations
  exportCases: (groupId: number) =>
    api.get(`/testcase/groups/${groupId}/export`, { responseType: 'blob' }),

  importCases: (groupId: number, file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return api.post<{ success: number; failed: number; errors: string[] }>(
      `/testcase/groups/${groupId}/import`,
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } }
    )
  }
}

export default testcaseApi
